<a href="https://colab.research.google.com/github/reguelpangrib29/prak-ai-2217051084-reguel-a/blob/main/AbsensiPertemuan/2217051084_Reguel_Andreas_Agustinus_Pangaribuan_AI_Linear_Classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Linear Classifier

## Studi Kasus

Membuat klasifikasi data kualitas Wine menggunakan model klasifikasi linear dengan akurasi minimal 60%

## Library

In [ ]:
import pandas as pd   # Mengimpor library pandas dengan alias pd untuk manipulasi data tabular.
import numpy as np    # Mengimpor library numpy dengan alias np untuk operasi numerik.
import sklearn        # Mengimpor library sklearn untuk machine learning.

# Mengimpor kelas LogisticReggresion dari modul linear_model di sklearn untuk membuat model regresi logistik.
from sklearn.linear_model import LogisticRegression

# Mengimpor fungsi train_test_split dari modul model_selection di sklearn untuk membagi data menjadi set latih dan set uji.
from sklearn.model_selection import train_test_split

# Mengimpor fungsi accuracy_score dari modul metrics di sklearn untuk mengukur akurasi model.
from sklearn.metrics import accuracy_score

## Dataset

In [ ]:
# Membaca data dari URL CSV ke dalam DataFrame menggunakan pandas.
data = pd.read_csv("https://raw.githubusercontent.com/adivii/datasets/main/Artificial%20Intelligence/Pertemuan%205/Wine_Quality_Data.csv")

data # Menampilkan DataFrame yang berisi data Wine Quality

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1138            6.3             0.510         0.13             2.3      0.076   
1139            6.8             0.620         0.08             1.9      0.068   
1140            6.2             0.600         0.08             2.0      0.090   
1141            5.9             0.550         0.10             2.2      0.062   
1142            5.9             0.645         0.12             2.0      0.075   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1138                 29.0                  40.0  0.99574  3.42       0.75   
1139                 28.0                  38.0  0.99651  3.42       0.82   
1140                 32.0                  44.0  0.99490  3.45       0.58   
1141                 39.0                  51.0  0.99512  3.52       0.76   
1142                 32.0                  44.0  0.99547  3.57       0.71   

      alcohol  quality    Id  
0         9.4        5     0  
1         9.8        5     1  
2         9.8        5     2  
3         9.8        6     3  
4         9.4        5     4  
...       ...      ...   ...  
1138     11.0        6  1592  
1139      9.5        6  1593  
1140     10.5        5  1594  
1141     11.2        6  1595  
1142     10.2        5  1597  

[1143 rows x 13 columns]

## Preprocessing dan Splitting Data

In [ ]:
# Import SimpleImputer
from sklearn.impute import SimpleImputer

# Penanganan data yang hilang (mengisi nilai yang hilang dengan nilai rata - rata)
imputer = SimpleImputer(strategy='mean')
imputer_data = imputer.fit_transform(data)

# Kembalikan data ke DataFrame
final_data = pd.DataFrame(imputer_data, columns=data.columns)

# Tampilkan hasil preprocessing
final_data

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1138            6.3             0.510         0.13             2.3      0.076   
1139            6.8             0.620         0.08             1.9      0.068   
1140            6.2             0.600         0.08             2.0      0.090   
1141            5.9             0.550         0.10             2.2      0.062   
1142            5.9             0.645         0.12             2.0      0.075   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1138                 29.0                  40.0  0.99574  3.42       0.75   
1139                 28.0                  38.0  0.99651  3.42       0.82   
1140                 32.0                  44.0  0.99490  3.45       0.58   
1141                 39.0                  51.0  0.99512  3.52       0.76   
1142                 32.0                  44.0  0.99547  3.57       0.71   

      alcohol  quality      Id  
0         9.4      5.0     0.0  
1         9.8      5.0     1.0  
2         9.8      5.0     2.0  
3         9.8      6.0     3.0  
4         9.4      5.0     4.0  
...       ...      ...     ...  
1138     11.0      6.0  1592.0  
1139      9.5      6.0  1593.0  
1140     10.5      5.0  1594.0  
1141     11.2      6.0  1595.0  
1142     10.2      5.0  1597.0  

[1143 rows x 13 columns]

In [ ]:
# final_data diatas sudah final, disini hanya memperlihatkan apa yang terjadi pada data setelah menggunakan scaling minmax

# import MinMaxScaler
from sklearn.preprocessing import MinMaxScaler

# Normalisasi fitur
normalizer = MinMaxScaler()
normalized_data = normalizer.fit_transform(imputer_data)

# Kembalikan data ke DataFrame
norm_data = pd.DataFrame(normalized_data, columns=data.columns)

# Tampilkan hasil preprocessing
norm_data

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0          0.247788          0.397260         0.00        0.068493   0.106845   
1          0.283186          0.520548         0.00        0.116438   0.143573   
2          0.283186          0.438356         0.04        0.095890   0.133556   
3          0.584071          0.109589         0.56        0.068493   0.105175   
4          0.247788          0.397260         0.00        0.068493   0.106845   
...             ...               ...          ...             ...        ...   
1138       0.150442          0.267123         0.13        0.095890   0.106845   
1139       0.194690          0.342466         0.08        0.068493   0.093489   
1140       0.141593          0.328767         0.08        0.075342   0.130217   
1141       0.115044          0.294521         0.10        0.089041   0.083472   
1142       0.115044          0.359589         0.12        0.075342   0.105175   

      free sulfur dioxide  total sulfur dioxide   density        pH  \
0                0.149254              0.098940  0.567548  0.606299   
1                0.358209              0.215548  0.494126  0.362205   
2                0.208955              0.169611  0.508811  0.409449   
3                0.238806              0.190813  0.582232  0.330709   
4                0.149254              0.098940  0.567548  0.606299   
...                   ...                   ...       ...       ...   
1138             0.417910              0.120141  0.416300  0.535433   
1139             0.402985              0.113074  0.472834  0.535433   
1140             0.462687              0.134276  0.354626  0.559055   
1141             0.567164              0.159011  0.370778  0.614173   
1142             0.462687              0.134276  0.396476  0.653543   

      sulphates   alcohol  quality        Id  
0      0.137725  0.153846      0.4  0.000000  
1      0.209581  0.215385      0.4  0.000626  
2      0.191617  0.215385      0.4  0.001252  
3      0.149701  0.215385      0.6  0.001879  
4      0.137725  0.153846      0.4  0.002505  
...         ...       ...      ...       ...  
1138   0.251497  0.400000      0.6  0.996869  
1139   0.293413  0.169231      0.6  0.997495  
1140   0.149701  0.323077      0.4  0.998121  
1141   0.257485  0.430769      0.6  0.998748  
1142   0.227545  0.276923      0.4  1.000000  

[1143 rows x 13 columns]

In [ ]:
# Memisahkan kolom target (quality) dan kolom fitur lainnya
X = final_data.drop("quality", axis=1)
y = final_data["quality"]

In [ ]:
# Normalisasi kolom fitur menggunakan MinMaxScaler
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

# Kolom target (quality) tidak di-normalisasi
y_categorical = pd.Series(pd.cut(y, bins=len(y.unique()), labels=False))

In [ ]:
# Memisahkan data menjadi data latih dan data uji
# X_train: fitur - fitur untuk data latih
# X_test: fitur - fitur untuk data uji
# y_train: target untuk data latih
# y_test: target untuk data uji
# test_size: proporsi data uji terhadap total data, disini diatur sebesar 20% (0.2)
# random_state: seed untuk mengontrol randomization. memastikan hasil yang reprodusibel
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y_categorical, test_size=0.2, random_state=42)

## Modelling and Prediction

In [ ]:
# Membuat model Logistic Regression
# max_iter: Jumlah iterasi maksimum yang digunakan oleh algoritma optimasi untuk menemukan parameter model.
# Nilai 7124 disini adalah contoh angka sembarang dan bukan angka mainstream.
# Anda dapat menyesuaikan nilai ini sesuai kebutuhan atau menggunakan nilai default jika tidak yakin.
model = LogisticRegression(max_iter=7124)

# Melatih model menggunakan data latih
model.fit(X_train, y_train)

LogisticRegression(max_iter=7124)

In [ ]:
#prediksi data test
y_pred = model.predict(X_test)

In [ ]:
# menghitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print("Akurasi model logistic regression: ", accuracy*100, "%")

Akurasi model logistic regression:  62.00873362445415 %


Dari perhitungan akurasi model logistic regression mencapai 62% atau sudah mencapai target studi kasus.
Namun kita tetap perlu evaluasi model jika perlu menaikkan akurasi prediksinya agar lebih tinggi.

## Comparing Models

In [ ]:
# Compare ke model Support Vector Machines
from sklearn.svm import SVC

# Deklarasi model SVM
svm_model = SVC()

# Melatih model SVM
svm_model.fit(X_train, y_train)

# Evaluasi akurasi model
svm_accuracy = svm_model.score(X_test, y_test)

print("Akurasi model SVM: ", svm_accuracy*100, "%")

Akurasi model SVM:  62.882096069869 %


Akurasi nya naik sedikit menjadi 62.9% menggunakan model Support Vector Machines

In [ ]:
# Compare ke model Naive Bayes
from sklearn.naive_bayes import GaussianNB

# Deklarasi model SVM
nb_model = GaussianNB()

# Melatih model SVM
nb_model.fit(X_train, y_train)

# Evaluasi akurasi model
nb_accuracy = nb_model.score(X_test, y_test)

print("Akurasi model Naive Bayes: ", nb_accuracy*100, "%")

Akurasi model Naive Bayes:  61.135371179039296 %


Menggunakan model Naive Bayes, akurasinya sedikit turun menjadi 61.1%